lib/spack/spack/util/web.py

```diff
@@ -85,6 +86,27 @@ def handle_starttag(self, tag, attrs):
                     self.links.append(val)
 
 
+def read_etag(url):
+    if isinstance(url, str):
+        remote_url = urllib.parse.urlparse(url)
+
+    if remote_url.scheme == "s3":
+        s3 = s3_util.get_s3_session(remote_url, method="push")
+        remote_path = remote_url.path
+        while remote_path.startswith("/"):
+            remote_path = remote_path[1:]
+        obj = s3.get_object(Bucket=remote_url.netloc, Key=remote_path)
+        if not obj:
+            raise FetchError("Could not obtain etag or s3 resource")
+        return obj["ETag"]
+    elif remote_url.scheme == "file":
+        return hashlib.md5(open(remote_url.path, "rb").read()).hexdigest()
```

Comment:
This is wrong. Use `spack.util.url.file_url_string_to_path`

Reply:
Also, avoid reading an entire file into memory; hashlib supports streaming

Reply:
I am just cribbing this from other places that use the `urlparse.path`, I am curious how that can be wrong.

As far as the hashlib streaming, I saw that but couldn't get it to work locally, it seems like it may only exist in a later versions? Or I am just doing it wrong.

```
>>> with open("/path/to/file", "rb") as f:
...     digest = hashlib.file_digest(f, "sha256")
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
AttributeError: module 'hashlib' has no attribute 'file_digest'
```

Reply:
Those other places got it wrong too, for example:

```
>>> spack.util.web.url_util.path_to_file_url("/hell world")
'file:///hell%20world'
```

and on Windows it's strictly required.

Reply:
See `spack.util.crypto.checksum` regarding chunks